### PR TITLE
[TTNN] Enforce SDPA tile alignment

### LINKED ISSUE #9268: SDPA fw/bw: enforce tile alignment on Q/K/V
> **Note: WIP testing of tt-metal uplift changes coverage.** Sub-issue of #9267.

New asserts at `sdpa_fw:72`, `sdpa_bw_q:83`, `sdpa_bw_kv:99`:

```
(!check_last_dim || logical[-1] == padded[-1]) && logical[-2] == padded[-2]
```

Q and K are checked on both sequence and head_dim; V on sequence only (unaligned vdim is supported via DiffVDim). The `[-2]` half is unconditional, so any S not a multiple of 32 fails.

No SDPA verifier in tt-mlir checks padding or alignment.

**This is not only a build break.** Per metal's comment, the softmax scaler is computed from the padded shape and the kernels iterate the padded sequence — so unaligned S is silently mis-scaling attention today. The uplift converts an existing numerics bug into a loud failure.

- [ ] Add the alignment check to `SDPAForwardOp::verify()` / `SDPABackwardOp::verify()` (TTIR + TTNN)
- [ ] Audit training models for S not a multiple of 32 — those have wrong numerics on current main

### PR DESCRIPTION
**Pull request by saiyambharara**
### Ticket

Fixes #9268.

### Problem description

The Metal SDPA operations require the sequence dimension of Q, K, and V to be tile-height aligned. They also require the head dimension of Q and K to be tile-width aligned. The head dimension of V is exempt because the operations support `DiffVDim`.

The TTNN SDPA verifiers did not enforce these requirements, allowing statically misaligned operations to pass compiler verification and fail later at runtime.

### What's changed

- Add a shared tile-alignment verifier for `ttnn.sdpa_fw` and `ttnn.sdpa_bw`.
- Require static Q, K, and V sequence dimensions to be multiples of `TILE_HEIGHT`.
- Require static Q and K head dimensions to be multiples of `TILE_WIDTH`.
- Leave dynamic dimensions for runtime validation.
- Preserve support for an unaligned V head dimension.
- Document the alignment requirements.
- Add direct verifier and composite promotion coverage for forward and backward SDPA.

#### Testing

- Selective `ttmlir-opt` build passed.
- 10 targeted SDPA `llvm-lit` tests passed.
- `clang-format --dry-run --Werror` passed.
- `git diff --check` passed.

### Checklist

- [x] New/Existing tests provide coverage for changes

### CODE REVIEW

File: include/ttmlir/Dialect/TTNN/IR/TTNNOps.td [hunk 1]
```diff
@@ -2830,8 +2830,14 @@ def TTNN_SDPAForwardOp : TTNN_Op<"sdpa_fw"> {
...
       backward pass.
 
-      All tensors are rank 4 `(B, H, S, D)`. `attention_mask` is only valid when
-      `mask_type` is `arbitrary`.
+      All tensors are rank 4 `(B, H, S, D)`. The sequence dimension `S` of Q,
+      K and V must be a multiple of `TILE_HEIGHT`, and the head dimension `D`
+      of Q and K must be a multiple of `TILE_WIDTH`. V's head dimension may be
+      unaligned because the backing operation supports a different V dimension
+      (`DiffVDim`). Static alignment is checked by the verifier; dynamic
+      dimensions remain subject to the same requirement at runtime.
```
Comment (amalkovTT) [inline include/ttmlir/Dialect/TTNN/IR/TTNNOps.td]: The last sentence is not needed. It is obvious that only static dimensions are checked. Also remove `DiffVDim` as that term is not used anywhere else in the project and the sentence before it is enough.

File: include/ttmlir/Dialect/TTNN/IR/TTNNOps.td [hunk 2]
```diff
@@ -2863,8 +2869,14 @@ def TTNN_SDPABackwardOp : TTNN_Op<"sdpa_bw"> {
...
       intermediates), computes the gradients w.r.t. query, key and value.
 
-      All tensors are rank 4 `(B, H, S, D)`. `attention_mask` is only valid when
-      `mask_type` is `arbitrary`.
+      All tensors are rank 4 `(B, H, S, D)`. The sequence dimension `S` of Q,
+      K and V must be a multiple of `TILE_HEIGHT`, and the head dimension `D`
+      of Q and K must be a multiple of `TILE_WIDTH`. V's head dimension may be
+      unaligned because the backing operation supports a different V dimension
+      (`DiffVDim`). Static alignment is checked by the verifier; dynamic
+      dimensions remain subject to the same requirement at runtime.
```
Comment (amalkovTT) [inline include/ttmlir/Dialect/TTNN/IR/TTNNOps.td]: Same here.

File: test/ttmlir/Dialect/TTIR/Decomposition/sdpa_bw_reshape.mlir
Comment (amalkovTT) [inline test/ttmlir/Dialect/TTIR/Decomposition/sdpa_bw_reshape.mlir:1]: This file is meant to test a transformation in TTIR. It is the wrong place to check if the TTNN operation is rejected. The tests that you added in TTNN are enough, so this file can be restored.

File: test/ttmlir/Dialect/TTNN/sdpa_fw/sdpa_fw_tests_negative.mlir [hunk 3]
```diff
@@ -0,0 +1,146 @@
+// SPDX-FileCopyrightText: (c) 2026 Tenstorrent AI ULC
+//
+// SPDX-License-Identifier: Apache-2.0
+
```
Comment (amalkovTT) [inline test/ttmlir/Dialect/TTNN/sdpa_fw/sdpa_fw_tests_negative.mlir]: We don't put copyright header in lit tests.

**Review by amalkovTT — COMMENTED** [review]
Thanks for the contribution! Few issues inline.

**Comment by saiyambharara** [comment]
Thanks for the review, @amalkovTT. I have addressed the comments and rerun the relevant tests. Could you please take another look and approve if everything looks good?

**Review by amalkovTT — APPROVED** [review]

**Comment by amalkovTT** [comment]
Merged. Thanks!
